### DISCUSSION TITLE
WebTransport?

### DISCUSSION THREAD

**Discussion by okbaswc** [microsoft/msquic#2078] [Q&A] [answered]
I had QuicTransport working with MSQUIC, but that has been dropped in favor of WebTransport.

What exactly is needed on top of MSQUIC to work with WebTransport? What is missing in the handshake etc. for MSQUIC to work with HTTP3?

**Comment by nibanks** [comment, marked as answer]
WebTransport is on top of HTTP/3 AFAIK. MsQuic is only the QUIC layer, not the HTTP pieces on top of that necessary for HTTP/3.

**Reply by okbaswc** [reply]
I guess my question is where exactly does MSQUIC fit into the WebTransport handshake? I guess I'd need to handle some HTTP3 frames, and then at some point pass things off to the port that MSQUIC is listening on (ListenerOpen(...)) callback?

**Reply by nibanks** [reply]
You will need to implement a minimal HTTP/3 on top of MsQuic. I've had other folks interested in this. If you'd be willing to do the work, we'd definitely take a contribution for an optional HTTP/3 layer/lib on top of MsQuic.

**Reply by okbaswc** [reply]
I'd be happy to do the work, just hard to find how to start. I fundamentally don't know where the Quic protocol comes into play with HTTP/3 and can't find any resources. Specs are too high level for me to go straight into coding from. Any tips? Doesn't Microsoft use MSQUIC in their HTTP/3 stack?

**Reply by nibanks** [reply]
Both Windows and .NET Core have an HTTP/3 implementation on top of MsQuic, but Windows' implementation is kernel mode and not open source, and .NET Core is C#, and both integrated into existing HTTP stacks.

For you, I'd recommend scoping down the problem to a minimal HTTP/3 solution; just enough to support WebTransport. Beyond the full-up [HTTP/3 doc](https://datatracker.ietf.org/doc/html/draft-ietf-quic-http), I don't have any great references that you might use. I believe [picoquic](https://github.com/private-octopus/picoquic) has a minimal HTTP/3 solution so that it can send and receive a GET request (used for interop testing). You might be able to reference some of that.
